File: Problem22.java
// Time Complexity : O(N)
// Space Complexity : O(1) we are editing given matrix
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this: Understood the problem
// Summary: go from bottom row and add minimum possible value until you get to top, then return min from top row


class Solution {
    public int minCost(int[][] costs) {
        if(costs == null || costs.length ==0){
            return 0;
        }
        for(int i = costs.length - 2; i>=0; i--){
            costs[i][2] = costs[i][2] + Math.min(costs[i+1][0], costs[i+1][1]);
            costs[i][1] = costs[i][1] + Math.min(costs[i+1][0], costs[i+1][2]);
            costs[i][0] = costs[i][0] + Math.min(costs[i+1][2], costs[i+1][1]);
        }
        
        return Math.min(costs[0][0], Math.min(costs[0][1], costs[0][2]));
        
    }
    
}

File: Problem_22.py
# Time Complexity : O(n) where n is the length of the list nums
# Space Complexity : Constant auxiliary space needed as we only use a few variables
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        n = len(costs)
        if not n:
            return
        rCost = costs[n-1][0] # Initialize rCost, bCost and cCost to their respective costs in the last row
        bCost = costs[n-1][1]
        gCost = costs[n-1][2]
        
        for i in range(n-2,-1,-1):
            # Iterate backwards from n-2 to 0
            tempR = rCost # Store in temp variables as their value will change in the middle of operations
            tempB = bCost
            tempG = gCost
            
            rCost = costs[i][0]+ min(tempB,tempG) # rCost will be cost of red paint for current house + min of Cost among Blue and Green paints for the next house
            bCost = costs[i][1]+ min(tempR,tempG)
            gCost = costs[i][2]+ min(tempR,tempB)
        
        return min(rCost,bCost,gCost) # Return whichever path leads to lowest cost
            